sklearn/decomposition/_dict_learning.py

```diff
@@ -1041,10 +1044,15 @@ def fit(self, X, y=None):
         self : object
             Returns the object itself
         """
+        self.components_ = self.dictionary
         return self
 
+    def transform(self, X, y=None):
+        self.fit(X, y)
+        return super().transform(X)
```

Comment:
This is really ugly. `transform` should not change the fit attributes. I don't see the point in change an estimator that does not follow the API into another that does not follow it either...

Instead we could have a private `_transform` method in `SparseCodingMixin` that would take the dictionary as argument and let the `SparseCodingMixin.transform` method call `SparseCodingMixin._transform` with `components_` by default but `SparseCoder.transform` would override it to pass `self.dictionary` instead.